### PR TITLE
Integrate api-check on the build

### PR DESCRIPTION
**Pull request by javiercn**

### CODE REVIEW

File: makefile.shade [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+#api-check target="verify"
+    @{
+        var Configuration = E("Configuration");
+        var BASE_DIR = Directory.GetCurrentDirectory();
+
+        var apiCheckFullFramework = new FileInfo(Path.Combine(BASE_DIR, ".build\\ApiCheck\\tools\\net452\\ApiCheck.exe"));
+        var apiCheckCore = new FileInfo(Path.Combine(BASE_DIR, ".build\\ApiCheck\\tools\\netcoreapp1.0\\ApiCheck.dll"));
+
+        var directories = Directory.EnumerateDirectories(Path.Combine(BASE_DIR, ".\\src"));
```
Comment (pranavkm) [inline makefile.shade]: `var srcProjects`

File: makefile.shade [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+        var fullFrameworkMonikerPrefix = new string[] { "net45", "net46" };
+        var coreMonikerPrefix = new string[] { "netcoreapp", "netstandard" };
+
+        foreach (var directoryPath in directories)
```
Comment (pranavkm) [inline makefile.shade]: `directoryPath` -> `srcProjectPath`

File: makefile.shade [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        {
+            var directoryInfo = new DirectoryInfo(directoryPath);
+            var projectJson = directoryInfo.EnumerateFiles().First(f => f.Name.Equals(PROJECT_JSON, StringComparison.OrdinalIgnoreCase));
```
Comment (pranavkm) [inline makefile.shade]: You should be able to `Path.Combine(directoryPath, "project.json");`. Alternate casings have difficulty building on xplat so we should have them all normalized.

File: makefile.shade [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    @{
+        var Configuration = E("Configuration");
+        var BASE_DIR = Directory.GetCurrentDirectory();
+
+        var apiCheckFullFramework = new FileInfo(Path.Combine(BASE_DIR, ".build\\ApiCheck\\tools\\net452\\ApiCheck.exe"));
+        var apiCheckCore = new FileInfo(Path.Combine(BASE_DIR, ".build\\ApiCheck\\tools\\netcoreapp1.0\\ApiCheck.dll"));
+
+        var directories = Directory.EnumerateDirectories(Path.Combine(BASE_DIR, ".\\src"));
+
+        var fullFrameworkMonikerPrefix = new string[] { "net45", "net46" };
```
Comment (pranavkm) [inline makefile.shade]: All of our repos should be `net451` on desktop. For `net46` of higher we would target the `netstandard` equivalent.

File: makefile.shade [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+                {
+                    var arguments = new List<string>
+                    {
+                        "compare",
+                        "-b " + fullFrameworkBaseline.FullName,
+                        "-a " + fullFrameworkDll.FullName,
+                        "-p " + projectJson.FullName,
+                        "-c " + Configuration
+                    };
+                    if(Environment.GetEnvironmentVariable("APICHECK_EPI") != null)
```
Comment (pranavkm) [inline makefile.shade]: what's EPI?

File: makefile.shade [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+        var Configuration = E("Configuration");
+        var BASE_DIR = Directory.GetCurrentDirectory();
+
+        var apiCheckFullFramework = new FileInfo(Path.Combine(BASE_DIR, ".build\\ApiCheck\\tools\\net452\\ApiCheck.exe"));
+        var apiCheckCore = new FileInfo(Path.Combine(BASE_DIR, ".build\\ApiCheck\\tools\\netcoreapp1.0\\ApiCheck.dll"));
+
+        var directories = Directory.EnumerateDirectories(Path.Combine(BASE_DIR, ".\\src"));
+
+        var fullFrameworkMonikerPrefix = new string[] { "net45", "net46" };
+        var coreMonikerPrefix = new string[] { "netcoreapp", "netstandard" };
```
Comment (pranavkm) [inline makefile.shade]: IMO, it should be fine to look for `netstandard`. The only `netcoreapp` targeting things we ship are tools and I don't think we'll want to guarantee API compatibility for those.

File: makefile.shade (hunk 5 above)
Comment (javiercn) [inline makefile.shade]: Exclude Public Internal

Comment (pranavkm) [inline makefile.shade]: 💩

**Comment by davidfowl** [comment]
This is just a test right? We're not actually going to put this into individual repositories right?

**Comment by javiercn** [comment]
@davidfowl yep, this will be part of KoreBuild but before enabling it for all repositories we are testing it on MVC and EF. Then we'll do all the other repositories

File: makefile.shade [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+                    }
+                }
+            }
+
+            if(fullFrameworkDll != null)
+            {
+                var fullFrameworkBaseline = directoryInfo.EnumerateFiles("baseline.net45.json").FirstOrDefault();
+
+                var generate = Environment.GetEnvironmentVariable("GENERATE_APICHECK_BASELINE");
+                if(!string.IsNullOrWhiteSpace(generate))
```
Comment (Eilon) [inline makefile.shade]: Check string.IsNullOrEmpty instead.

File: makefile.shade (hunk 5 above)
Comment (Eilon) [inline makefile.shade]: Add a doc to this or use a different env var name. The latter is preferred.

File: makefile.shade [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+                {
+                    coreDll = coreDirectory.EnumerateFiles(string.Concat(directoryInfo.Name,".dll")).FirstOrDefault();
+                    if(fullFrameworkDll != null)
+                    {
+                        break;
+                    }
+                }
+            }
+
+            if(coreDll != null)
```
Comment (Eilon) [inline makefile.shade]: Can this all be in a helper method? Everything is duplicated here...

**Comment by Eilon** [comment]
@davidfowl this is part of our breaking change detection. It is intended to run on every build.

File: makefile.shade (hunk 4 above)
Comment (javiercn) [inline makefile.shade]: I know, but let's be better safe than sorry. We will have to migrate this to MSBuild, so I'm not very worried about it.

File: makefile.shade (hunk 6 above)
Comment (javiercn) [inline makefile.shade]: Again, better safe than sorry.

**Comment by javiercn** [comment]
Updated

File: makefile.shade (hunk 4 above)
Comment (pranavkm) [inline makefile.shade]: 👎

File: makefile.shade [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+functions @{
+
+    string GenerateApiCheckInvoker(string framework)
+    {
+        var BASE_DIR = Directory.GetCurrentDirectory();
+        var apiCheckFullFramework = new FileInfo(Path.Combine(BASE_DIR, ".build\\ApiCheck\\tools\\net452\\ApiCheck.exe"));
+
+        if(framework == "net45")
```
Comment (pranavkm) [inline makefile.shade]: Bracing style is inconsistent. Can you add a space after all the `if`s?

File: makefile.shade [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+                    arguments.Add("-e " + apiListingExceptions.FullName);
+                }
+
+                if(Environment.GetEnvironmentVariable("APICHECK_EXCLUDE_PUBLIC_INTERNAL_TYPES") != null)
+                {
+                    // Excludes types in the internal namespace
+                    arguments.Add("-epi");
+                }
+
+                Exec(GenerateApiCheckInvoker(framework), string.Join(" ",arguments));
```
Comment (pranavkm) [inline makefile.shade]: Formatting - `string.Join(" ", arguments)`

**Pull request closed without merging**